A lot of changes

RaphaelIT7 opened the pull request:
Apparently renaming a branch closes a pr, so here is another one but with the right branch.

This is a draft, as I still want to get some things done, like finishing the ambient sounds.
Anyone can already review it/request changes, but right now I don't see this as being ready to be merged.
NOTE: I'm going to add a proper todo list of the things left I want to do later.

I mainly created this draft already, as there was interest in if there would be a pr for these changes.

## Note
All newly added sounds will be from the SlashCo VR version.

## gameplay changes
[+] A dark fog that dynamically adjusts to the environment lighting.
[+] Added a new item "Tesla Coil"
(It stuns all slashers on the entire map for 20 seconds)
[+] Added a clientside sound that is played when a Slasher sees a survivor the first time.
[+] Added basic slasher documents
(This is still WIP and is unfinished)
[+] Added a status light to the generators showing if their running or not
[+] Added a fuel display to the generators allowing one to see how much fuel is missing
[+] Implemented quick escape reward (10 credits)
[+] Different game intro sounds are played based off the slasher difficulty.
[+] Playing the dangerLevel sound when spawning into a round
[+] Implemented quick escape and slow escape.
(If you escape before the round went on for 10 minutes it counts as a quick escape, if you escape after 20 minutes its a slow escape)
[+] Added ambient sounds which volume and sound track are based off the Slasher's anger.
[#] Disable broken door collisions after it stopped moving
(if you accidentally walked on it previously, your movement would become jittery as the prediction would fail).
[#] Allow one to drop the used deathward

### Tyler
[+] Added a camera shake to his effect, intensity is based off his distance
[+] Added new sound for when he's below 50 anger
[+] Implement endless chase if the time is up
(when the round time reaches the defined slow time of 20min, this chase is disabled by default)
[+] If a player is holding a fuel can, Tyler will destroy the fuel can first instead of killing the player
[+] Added Tyler theme which is played as background music after the first chase.
[+] Added three new songs from SlashCo VR to play when hes tyler the creator.
[#] Made his HUD effect local and based off his distance to the player
[#] Move all sounds into `slashco/slasher/igor`
[#] Switched all code over to use the new audiosystem and the new anger system

### The Watcher
[+] Added footstep sounds from SlashCo VR
[+] Added background music from SlashCo VR
[#] Move all sounds into `slashco/slasher/watcher`

### Princess
[+] If a player is holding a baby, Princess will eat the baby now.
[#] When Princess eats/kills a player or baby, they will be stunned for 4.5 seconds.

### Baby Item
[#] A random slasher is now teleported instead of always the second one.

## Other changes
[+] Precached a lot of things to reduce in game laggs.
[+] Precached the next map
(The download of the next map starts in the lobby as soon as it is selected to reduce loading times later on).
[+] Added missing clientside prediction for Impervious
(going through doors/players as male07 is now far smoother).
[+] Spectators have an animation and watch the helicopter take off in the lobby.
[+] Added missing clientside prediction for Impervious
(going through doors/players as male07 is now far smoother).
[+] Implemented a new audiosystem for lobby and future ambient sounds
[+] Support live language changes
[+] Added sound/vision fade in when spawning into a round
[#] Implemented a failsafe in case the slashers or survivors disconnect
(the game ends if no survivors or slashers exist after 90 seconds)
[#] Solved a spectator prediction issue, making movement/noclipping jittery
(It was a gmod bug)
[#] Stopped player from suiciding when the helicopter is taking off in the lobby.
[#] The picked slasher in the lobby can freely spectate instead of being locked in the spectator camera.
[#] solved an error caused by entities being created too early.
[#] cleaned up the entire code.
[#] Fixed auto refresh errors/bugs that made development more painful.
[#] Fixed lobby-ready status clipping into player count
[#] Fixed spectators being able to leave noclip
[#] Rendering difficulty colors in the round info screen now
[#] Changed helicopter escape music to use the new audio system
[#] Fixed a game crash caused by a infinite loop if no map was found
[#] Moved a bit away from NW to improve networking and possibly support replays/demos.
[#] Converted almost all .wav files into .mp3 and into mono to solve some 3d sound issues (might also solve #29)
[#] Fixed player model selector being always broken(fixes #33)
[#] Fixed round end panels stacking and never removing the old panel
[#] Fixed helicopter collisions being left behind
(It's physics object didn't follow the helicopter causing a invisible box with collisions to exist)
[#] Syncronized Helicopter voice lines for all players
[#] Fixed a possible error with princess
[#] Live update points and wins when their changed
[#] Show the ready state of other players in the lobby (if they picked slasher or survivor)
[#] Show the spectator ui to the slasher when their waiting to be spawned.
[#] Fixed `gmod_hands` sometimes spawning and causing errors
[#] Fixed a bug with particle emitters used for footsteps and decoy causing issues when they don't clean up quick enouth & optimized the code of it to only use a single emitter now.
[-] Removed hardcoded 7-player limit.
(If the helicopter is full, additional players will be somewhat bugged, but it should work fine.)

## Lua API Changes
This documentation wasn't finished yet

[+] Added `SlashCo.States`, `SlashCo.DifficultyLevel`, `SlashCo.SlasherClass`, `SlashCo.DangerLevel` enums
[+] Added `SlashCo.GetDangerColor`, `SlashCo.GetDangerSound`, `SlashCo.GetNameColor`, `SlashCo.GetClassColor` functions
[#] Changed `SlashCo.OfferingData` keys.
\-> Renamed `SO` to `Singularity`
\-> Renamed `DO` to `Duality`
\-> Renamed `SatO` to `Satiation`

RaphaelIT7 committed:
autorun: use a single file for playermodels

RaphaelIT7 committed:
gamemode: change network strings to make them a bit more readable

RaphaelIT7 committed:
gamemode: formatting changes & more

RaphaelIT7 committed:
repo: implement shared GameData table for commonly used vars

RaphaelIT7 committed:
gamemode: fix lobby hud breaking on autorefresh

RaphaelIT7 committed:
slashers: try to improve prediction?

RaphaelIT7 committed:
gamemode: do not use magic values

RaphaelIT7 committed:
gamemode: disable destroyed door collision once it stopped moving

RaphaelIT7 committed:
gamemode: have helicopter start animation for spectators

RaphaelIT7 committed:
gamemode: implement map fog

RaphaelIT7 committed:
gamemode: disable halo around dormant entities

RaphaelIT7 committed:
gamemode: fix autorefresh breaking state, remove hardcoded 7 player limit, change formatting

RaphaelIT7 committed:
gamemode: add the current map to server download

RaphaelIT7 committed:
gamemode: remove a debug print

RaphaelIT7 committed:
gamemode: precache the next level to reduce loading times

RaphaelIT7 committed:
gamemode: solved some errors caused by too early entity creation

RaphaelIT7 committed:
gamemode: fix error when precaching the next map

RaphaelIT7 committed:
gamemode: remove a debug print

RaphaelIT7 committed:
repo: do some formatting

RaphaelIT7 committed:
gamemode: allow slasher to spectate freely in lobby instead of being stuck waiting

RaphaelIT7 committed:
gamemode: add clientside prediction to PLAYER:SetImpervious

RaphaelIT7 committed:
gamemode: stop player suicide when the helicopter is taking off

RaphaelIT7 committed:
gamemode: solve spectator prediction issues

RaphaelIT7 committed:
entities: remove the additional AddCSLuaFile calls

RaphaelIT7 committed:
repo: precache all models properly

RaphaelIT7 committed:
gamemode: create DifficultyLevel enum

RaphaelIT7 committed:
gamemode: implement failsafe in case slasher or survivers leave/never join

RaphaelIT7 committed:
repo: precache more things

RaphaelIT7 committed:
gamemode: remove useless spaces

RaphaelIT7 committed:
gamemode: add missing precache call

RaphaelIT7 committed:
gamemode: fix a error

RaphaelIT7 committed:
gamemode: fix a error caused by difficulty level changes

RaphaelIT7 committed:
gamemode: fix precache being called too early

RaphaelIT7 committed:
materials: making some content smaller

RaphaelIT7 committed:
gamemode: fix a client error

RaphaelIT7 committed:
gamemode: always network survivors and slashers

RaphaelIT7 committed:
models: remove unused files

RaphaelIT7 committed:
repo: setup unlisted workshop for the new content

RaphaelIT7 committed:
gamemode: add a new item (Tesla Coil)

RaphaelIT7 committed:
gamemode: disable fog in the lobby & for spectators

RaphaelIT7 committed:
gamemode: add a refresh counter because why not

RaphaelIT7 committed:
repo: use tabs instead of spaces for more consistency

RaphaelIT7 committed:
gamemode: disable fog for slashers

FirstLemon committed:
updated german localization (#1)

RaphaelIT7 committed:
ui: solve ready status clipping into player count

RaphaelIT7 committed:
audiosystem: implement & use a new audiosystem for background music

RaphaelIT7 committed:
gamemode: stop spectators from leaving noclip

RaphaelIT7 committed:
gamemode: disable background music in offerings

RaphaelIT7 committed:
repo: a bit more cleanup

RaphaelIT7 committed:
gamemode: render colors in round info screen

RaphaelIT7 committed:
gamemode: play a sound when a survivor has been seen by a slasher

RaphaelIT7 committed:
items: increase tesla coil price to 50

RaphaelIT7 committed:
gamemode: remove the refresh counter as it shouldn't be here

RaphaelIT7 committed:
gamemode: put debug messages behind a debug var to not fill the console

RaphaelIT7 committed:
gamemode: change back workshop id & remove this stuff for the pr

RaphaelIT7 commented:
If wanted I could split this PR into multiple smaller ones

RaphaelIT7 committed:
gamemode: readd removed proximity code

RaphaelIT7 committed:
gamemode: add slashco_proximity_range convar & small changes

RaphaelIT7 committed:
gamemode: add slashco_unstuck command

RaphaelIT7 committed:
gamemode: move ready status a bit again

RaphaelIT7 committed:
gamemode: add volume support for background music & implement basic anger system

RaphaelIT7 committed:
lang: update description to match current stun time

RaphaelIT7 committed:
gamemode: solve two issues

RaphaelIT7 committed:
gamemode: handle content changes

RaphaelIT7 committed:
gamemode: make name argument optional for register item/effect/slasher functions

RaphaelIT7 committed:
gamemode: small changes

RaphaelIT7 committed:
gamemode: add basic document system (basic functionality & documents)

RaphaelIT7 committed:
helicopter: improve clientside FPS

RaphaelIT7 committed:
gamemode: move lobby failsafe code into the right file

RaphaelIT7 committed:
audiosystem: solve it skipping back in singleplayer.

RaphaelIT7 committed:
gamemode: implement quick escape code

RaphaelIT7 committed:
ui: adjust document ui to match better

RaphaelIT7 committed:
gamemode: use audiosystem for helicopter escape music

RaphaelIT7 committed:
gamemode: fix a error caused by previous rename of the function

RaphaelIT7 committed:
gamemode: improve FindPlayersInView code & support slasher's visibility function

RaphaelIT7 committed:
gamemode: fix a infinite loop & unstuck falsely reporting & two other small changes

textstack commented:
> If wanted I could split this PR into multiple smaller ones

sorry for not responding sooner, totally fine with one big pr and the only thing i see that's worth contesting is the tesla coil (we've already talked about it a little)

RaphaelIT7 commented:
> > If wanted I could split this PR into multiple smaller ones
> 
> sorry for not responding sooner, totally fine with one big pr and the only thing i see that's worth contesting is the tesla coil (we've already talked about it a little)

I did add the Tesla coil mostly because I was bored in the middle of the night and thought, Why not make an item xd
So it definitely has some flaws and/or points to contest.
Also, I never really played with a huge number of players (mostly test stuff locally with multiple gmod instances as I don't get that many people to play with), so I also have no idea how it affects the gameplay in bigger lobbies.

A small sidenote so that the direction of this PR is a bit more clear:
I try with this PR to bring some things over from SlashCo VR, as I do love playing the VrC version, and things like the document and perk systems or even things like the slasher anger system and such would be my goal to implement here.
I'll also do any bug fixes, cleanups or improvements I find along the way, but other than that, there isn't any other goal right now.

RaphaelIT7 committed:
gamemode: use the proper collisiongroup when checking if a player is stuck + two other small changes

RaphaelIT7 committed:
genetator: add fuel display & add a status light

textstack commented on gamemodes/slashco/entities/entities/sc_generator.lua:
```diff
@@ -19,23 +19,88 @@ hook.Add("SlashCo:Precache", "PrecacheGenerator", function()
 		self:DrawModel()
+
+		-- Small fuel UI showing how full a generator is
+		local maximumFuel = 4 -- How many fuel canisters can go into a single generator, if this changes this code needs to be updated manually.
```
Is there something stopping you from being able to use the global int for gas cans per generator? That can actually be changed on a per-map level by mapmakers.

RaphaelIT7 replied:
In the current implementation it would break the UI, I had planned to change this later on/have it on my ToDo.
It'll be one of the things I'll do today.

textstack replied:
cool

RaphaelIT7 replied:
![image](https://github.com/user-attachments/assets/d01b3014-4695-4e6b-97ea-047180a90122)
:D

RaphaelIT7 committed:
generator: don't hard code the maximum fuel count

RaphaelIT7 committed:
generator: ensure that it never breaks in case the remaining value gets screwed up.

RaphaelIT7 committed:
gamemode: don't hardcode the lobby.

RaphaelIT7 committed:
gamemode: save the last lobby that was played on

RaphaelIT7 committed:
gamemode: network GameData.Lobby just so that the client has if it'll ever be needed.

RaphaelIT7 committed:
repo: move a bit away from NW

RaphaelIT7 committed:
repo: use enum's for offerings instead of magic numbers

RaphaelIT7 committed:
entities: small cleanup

RaphaelIT7 committed:
gamemode: implement quick escape reward

textstack commented on gamemodes/slashco/gamemode/sv_globals.lua:
```diff
@@ -66,7 +66,7 @@ function SlashCo.ResetCurRoundData()
 			GasCanMod = 0,
-			SO = 0,
+			SO = 0, -- ToDo: rename this and the two below, what do they do, what are they? no idea :(
 			DO = false,
```
i think singularity and duality offering? the offering system in general is somewhat rotten

RaphaelIT7 committed:
gamemode: use our enums

RaphaelIT7 committed:
gamemode: disable background music serverside on gameintro

RaphaelIT7 committed:
gamemode: play different intro sound based off difficulty

RaphaelIT7 committed:
gamemode: remove this useless line

RaphaelIT7 committed:
gamemode: change offering table keys

RaphaelIT7 committed:
tyler: make his effect not global & add a camera shake

RaphaelIT7 committed:
gamemode: implement fog for slashers based on eyesight

RaphaelIT7 committed:
gamemode: support live language changes

RaphaelIT7 committed:
lang: fix translation key for quickescape (was formerly named points_fast)

RaphaelIT7 committed:
gamemode: add SlashCo.AddDangerLevel, SlashCo.AddSlasherClass and SlashCo.AddOffering functions

RaphaelIT7 committed:
gamemode: implement backwards compatibility with any addons that might have broken

RaphaelIT7 committed:
gamemode: shorten this

RaphaelIT7 committed:
gamemode: improve audiosystem

RaphaelIT7 committed:
repo: convert almost all .wav to .mp3

RaphaelIT7 committed:
audiosystem: solve a bug which caused the background music to play multiple times

RaphaelIT7 committed:
gamemode: properly support multiple multirun instances at once

RaphaelIT7 committed:
playermodelpicker: fix it always being broken

https://github.com/Mantibro/SlashCo/issues/33

RaphaelIT7 committed:
gamemode: change survivor model code a bit

RaphaelIT7 committed:
gamemode: use proper enums

RaphaelIT7 committed:
princess: solve a possible error

RaphaelIT7 committed:
gamemode: change helicopter code a bit

RaphaelIT7 committed:
gamemode: syncronize helicopter voice for all clients

RaphaelIT7 committed:
gamemode: forgot to change the enum here too

RaphaelIT7 committed:
helicopter: fix it's physics object being left behind

RaphaelIT7 committed:
gamemode: fixed a few things & added screen/audio fade when landing

RaphaelIT7 committed:
gamemode: revert this

RaphaelIT7 committed:
audiosystem: implement precache logic & use SlashCo VR's terminal sounds for document ui

RaphaelIT7 committed:
audiosystem: finish playsound & stopsound implementations

RaphaelIT7 committed:
gamemode: add a missing null check

RaphaelIT7 committed:
tyler: change a lot of things

[+] Added new sound for when he's below 50 anger
[+] Implement endless chase if the time is up
(when the round time reaches the defined slow time of 20min, this chase is disabled by default)
[+] If a player is holding a fuel can, Tyler will destroy the fuel can first instead of killing the player
[+] Added Tyler theme which is played as background music after the first chase.
[#] Move all sounds into slasher/igor
[#] Switch to use the new audiosystem
[#] Switch to use the new anger system

The new songs are all from SlashCo VR

RaphaelIT7 committed:
tyler: add three new songs

RaphaelIT7 committed:
watcher: implement new sounds & implement background music

RaphaelIT7 committed:
repo: do a lot of bug fixes

RaphaelIT7 committed:
helicopter: hopefully solve the helicopter sounds continuing to play

RaphaelIT7 committed:
repo: use MAX_EDICT_BITS for EntIndex networking

RaphaelIT7 committed:
tyler: remove unessesary code

RaphaelIT7 committed:
repo: rework audiosystem PlaySound function

RaphaelIT7 committed:
gamemode: fix documents not being assigned for dead survivors

RaphaelIT7 commented:
I put my entire ToDo that I had locally into here for more transparency on this PR: https://github.com/users/RaphaelIT7/projects/6

RaphaelIT7 committed:
gamemode: Show the ready state of others & optimize this code a bit

RaphaelIT7 committed:
gamemode: Show slasher the spectator ui

fixes https://github.com/RaphaelIT7/SlashCo/issues/36

RaphaelIT7 committed:
gamemode: Allow one to drop the used deathward

fixes #41

RaphaelIT7 committed:
gamemode: live update points and wins

fixes #35

RaphaelIT7 committed:
gamemode: implement slow escape & time warning

fixes https://github.com/RaphaelIT7/SlashCo/issues/4

RaphaelIT7 committed:
slasher: use the right function here

RaphaelIT7 committed:
gamemode: implement ambient tracks based of the highest slasher anger

RaphaelIT7 committed:
princess: eat the baby if a player is holding one

fixes https://github.com/RaphaelIT7/SlashCo/issues/37

RaphaelIT7 committed:
gamemode: add a safety check to PLAYER:SetItem for the slot argument

RaphaelIT7 committed:
audiosystem: allow one to stop sounds globally.

fixes https://github.com/RaphaelIT7/SlashCo/issues/27

RaphaelIT7 committed:
gamemode: do a lot of bug fixes

[#] Fixed some potential lua errors with emitters & improved the performance a bit
[#] Actually get the slasher spectator hud to show right
[#] Fix princess not always eating the baby & freeze princess while eating a baby or player
[#] Princess should not damage the player that is holding the player
[#] Fix lobby failsafe kicking in when debugging
[#] Fix gmod_hands somehow being created and causing warnings to be spammed
[#] Raise Helicopter sound levels.

RaphaelIT7 committed:
gamemode: don't run these timers if were in the lobby
